include/cpp2util.h

```diff
@@ -1629,39 +1629,44 @@ constexpr auto is( X&& ) {
 //  Types
 //
 template< typename C, typename X >
-auto is( X const& x ) -> bool {
+constexpr auto is( X const& x ) -> auto {
     if constexpr (
         std::is_same_v<C, X>
         || std::is_base_of_v<C, X>
     )
     {
-        return true;
+        return std::true_type{};
     }
     else if constexpr (
-        std::is_base_of_v<X, C>
-        || (
-            std::is_polymorphic_v<C>
-            && std::is_polymorphic_v<X>
-            )
+        std::is_polymorphic_v<C>
+        && std::is_polymorphic_v<X>
     )
     {
-        if constexpr (std::is_pointer_v<X>) {
-            return Dynamic_cast<C const*>(x) != nullptr;
-        }
-        else {
-            return Dynamic_cast<C const*>(&x) != nullptr;
-        }
+        return Dynamic_cast<C const*>(&x) != nullptr;
     }
     else if constexpr (
-        requires { *x; X(); }
+        (
+            std::is_same_v<X, std::nullptr_t>
+            || requires { *x; X(); }
+        )
         && std::is_same_v<C, empty>
     )
     {
         return x == X();
     }
-    else {
+    else if constexpr (
+        std::is_pointer_v<C>
+        && std::is_pointer_v<X>
+    )
+    {
+        if (x != nullptr) {
+            return bool{is<std::remove_pointer_t<C>>(*x)};
+        }
         return false;
     }
+    else {
+        return std::false_type{};
```

Comment:
As above: Can you summarize the benefit of `false_type` over `bool`?

Also, I haven't fully checked, just asking: Are the choices of `[true|false]_type` vs `bool` paired so that the former or the latter is consistently used for a given kind of successful/failed test?

Reply:
I believe that `true_type/false_type` means that the result is known at compile time because it's based solely on the types involved, and that `true/false` means that it's deferred to runtime because it depends on the value of the variable.

Reply:
Using `false_type`/`true_type` over `bool` gives the possibility of using the result in the compile-time. So, the result can be used in the `if constexpr`s or `requires` clause.

That means we know the result of some `is()` calls at compile time, which is useful for inspecting the types.